Mathlib/Combinatorics/SetFamily/Shatter.lean

```diff
@@ -0,0 +1,195 @@
+Authors: Yaël Dillies
+-/
+import Mathlib.Data.Finset.Sort
+import Mathlib.Data.Nat.Interval
+import Mathlib.Order.UpperLower.Basic
+import Mathlib.Combinatorics.SetFamily.Compression.Down
+
+/-!
+# Shattering families
+
+This file defines the shattering property and VC-dimension of set families.
+
+## Main declarations
+
+* `Finset.Shatter`: The shattering property.
+* `Finset.shatterer`: The set family of sets shattered by a set family.
+* `Finset.vcDim`: The Vapnik-Chervonenkis dimension.
+
+## TODO
+
+* Order-shattering
+* Strong shattering
+-/
+
+open scoped BigOperators FinsetFamily
+
+namespace Finset
+variable {α : Type*} [DecidableEq α] {𝒜 ℬ : Finset (Finset α)} {s t : Finset α} {a : α} {n : ℕ}
+
+/-- A set family shatters a set `s` if all subsets of `s` can be obtained as the intersection of `s`
+and some element of the set family. We also say that `s` is traced by `𝒜`. -/
+def Shatter (𝒜 : Finset (Finset α)) (s : Finset α) : Prop := ∀ ⦃t⦄, t ⊆ s → ∃ u ∈ 𝒜, s ∩ u = t
+
+instance : DecidablePred 𝒜.Shatter := fun _s ↦ decidableForallOfDecidableSubsets
+
+lemma Shatter.exists_inter_eq_singleton (hs : Shatter 𝒜 s) (ha : a ∈ s) : ∃ t ∈ 𝒜, s ∩ t = {a} :=
+  hs $ singleton_subset_iff.2 ha
+
+lemma Shatter.mono_left (h : 𝒜 ⊆ ℬ) (h𝒜 : 𝒜.Shatter s) : ℬ.Shatter s :=
+  fun _t ht ↦ let ⟨u, hu, hut⟩ := h𝒜 ht; ⟨u, h hu, hut⟩
+
+lemma Shatter.mono_right (h : t ⊆ s) (hs : 𝒜.Shatter s) : 𝒜.Shatter t := fun u hu ↦ by
+  obtain ⟨v, hv, rfl⟩ := hs (hu.trans h); exact ⟨v, hv, inf_congr_right hu $ inf_le_of_left_le h⟩
+
+lemma Shatter.exists_superset (h : 𝒜.Shatter s) : ∃ t ∈ 𝒜, s ⊆ t :=
+  let ⟨t, ht, hst⟩ := h Subset.rfl; ⟨t, ht, inter_eq_left.1 hst⟩
+
+lemma shatter_of_forall_subset (h : ∀ t, t ⊆ s → t ∈ 𝒜) : 𝒜.Shatter s :=
+  fun t ht ↦ ⟨t, h _ ht, inter_eq_right.2 ht⟩
+
+protected lemma Shatter.nonempty (h : 𝒜.Shatter s) : 𝒜.Nonempty :=
+  let ⟨t, ht, _⟩ := h Subset.rfl; ⟨t, ht⟩
+
+@[simp] lemma shatter_empty : 𝒜.Shatter ∅ ↔ 𝒜.Nonempty :=
+  ⟨Shatter.nonempty, fun ⟨s, hs⟩ t ht ↦ ⟨s, hs, by rwa [empty_inter, eq_comm, ←subset_empty]⟩⟩
+
+protected lemma Shatter.subset_iff (h : 𝒜.Shatter s) : t ⊆ s ↔ ∃ u ∈ 𝒜, s ∩ u = t :=
+  ⟨fun ht ↦ h ht, by rintro ⟨u, _, rfl⟩; exact inter_subset_left _ _⟩
+
+lemma shatter_iff : 𝒜.Shatter s ↔ 𝒜.image (fun t ↦ s ∩ t) = s.powerset :=
+  ⟨fun h ↦ by ext t; rw [mem_image, mem_powerset, h.subset_iff],
+    fun h t ht ↦ by rwa [←mem_powerset, ←h, mem_image] at ht⟩
+
+lemma univ_shatter [Fintype α] : univ.Shatter s := shatter_of_forall_subset $ fun _ _ ↦ mem_univ _
+
+@[simp] lemma shatter_univ [Fintype α] : 𝒜.Shatter univ ↔ 𝒜 = univ := by
+  rw [shatter_iff, powerset_univ]; simp_rw [univ_inter, image_id']
+
+/-- The set family of sets that are shattered by `𝒜`. -/
+def shatterer (𝒜 : Finset (Finset α)) : Finset (Finset α) := (𝒜.biUnion powerset).filter 𝒜.Shatter
+
+@[simp] lemma mem_shatterer : s ∈ 𝒜.shatterer ↔ 𝒜.Shatter s := by
+  refine mem_filter.trans $ and_iff_right_of_imp $ fun h ↦ ?_
```

Comment:
```suggestion
  refine mem_filter.trans <| and_iff_right_of_imp <| fun h ↦ ?_
```
I'm pretty sure `<|` is preferred over `$`, if I'm wrong feel free to disregard these suggestions